snapcraft/plugins/_python/_pip.py

```diff
@@ -0,0 +1,263 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2016-2017 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import collections
+import contextlib
+import json
+import logging
+import os
+import shutil
+import subprocess
+import sys
+import tempfile
+
+import snapcraft
+from snapcraft import file_utils
+from ._python_finder import (
+    get_python_command,
+    get_python_headers,
+)
+from . import errors
+
+logger = logging.getLogger(__name__)
+
+
+class Pip:
+    """Wrapper for pip abstracting the nasty args necessary for use in a part.
+
+    This class takes care of fetching pip, setuptools, and wheel, and then
+    simply shells out to pip with the magical arguments necessary to install
+    packages into a part. The important functions are:
+
+    download(): Fetch the provided list of packages (but do not install them)
+    install(): Install the provided list of packages (but do not fetch them)
+    wheel(): Build wheels for the provided list of packages (but do not fetch
+             them)
+    list(): List the installed packages
+    clean_packages(): Remove the package cache
+    env(): Obtain the environment required when calling pip
+
+    Of particular importance: packages must be downloaded (via download())
+    before they can be installed or have wheels built.
```

Comment:
Yeah the API seems error-prone, I agree. Especially because I'm not sure how easy it would be to accurately detect if one is trying to install a package that has not been downloaded. However, it's done this way because the download should only be done in the pull step, but the install needs to happen in both pull and build.